File: running_app/lib/models/activity/user_participation.dart

abstract class UserParticipation {
  final String? userId;
  final bool? isAdmin;
  final String? participatedAt;

  UserParticipation({
    required this.userId,
    required this.isAdmin,
    required this.participatedAt
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'is_admin': isAdmin,
      'participated_at': participatedAt,
    };
  }

  @override
  String toString() {
    return '${toJson()}';
  }
}

class UserParticipationClub extends UserParticipation {
  final String? clubId;

  UserParticipationClub({
    String? userId,
    bool? isAdmin,
    String? participatedAt,
    required this.clubId
  }) : super(
    userId: userId,
    isAdmin: isAdmin,
    participatedAt: participatedAt
  );

  // UserParticipationClub.fromJson(Map<String, dynamic> json)
  //     : club = CreateClub.fromJson(json['club']),
  //       super(
  //         user: User.fromJson(json['user']),
  //         isAdmin: json['is_admin'],
  //         participatedAt: json['participated_at'],
  //       );

  @override
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = super.toJson();
    data['club_id'] = clubId;
    return data;
  }

  @override
  String toString() {
    return 'UserParticipationClub{${super.toString()}, club: $clubId}';
  }
}

class UserParticipationEvent extends UserParticipation {
  final String? eventId;
  final bool? isSuperAdmin;

  UserParticipationEvent({
    String? userId,
    bool? isAdmin,
    String? participatedAt,
    required this.isSuperAdmin,
    required this.eventId
  }) : super(
      userId: userId,
      isAdmin: isAdmin,
      participatedAt: participatedAt
  );

  // UserParticipationEvent.fromJson(Map<String, dynamic> json)
  //     : event = Event.fromJson(json['event']),
  //       super(
  //         user: User.fromJson(json['user']),
  //         isAdmin: json['is_admin'],
  //         participatedAt: json['participated_at'],
  //       );

  @override
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = super.toJson();
    data['event_id'] = eventId;
    data['is_superadmin'] = isSuperAdmin;
    return data;
  }

  @override
  String toString() {
    return '${super.toJson()} ${toJson()}';
  }
}

class UserParticipationGroup extends UserParticipation {
  final String? groupId;

  UserParticipationGroup({
    String? userId,
    bool? isAdmin,
    String? participatedAt,
    required this.groupId
  }) : super(
      userId: userId,
      isAdmin: isAdmin,
      participatedAt: participatedAt
  );

  // UserParticipationGroup.fromJson(Map<String, dynamic> json)
  //     : group = CreateGroup.fromJson(json['group']),
  //       super(
  //         user: User.fromJson(json['user']),
  //         isAdmin: json['is_admin'],
  //         participatedAt: json['participated_at'],
  //       );

  @override
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = super.toJson();
    data['group_id'] = groupId;
    return data;
  }

  @override
  String toString() {
    return 'UserParticipationGroup{${super.toString()}, group: $groupId}';
  }
}

File: running-app-backend/activity/serializers/user_participation.py
from rest_framework import serializers

from activity.models import UserParticipationClub, \
                            UserParticipationEvent, \
                            UserParticipationGroup  

from account.serializers import UserSerializer
from activity.serializers import ClubSerializer, \
                                EventSerializer, \
                                GroupSerializer
                                
# from account.serializers import ActivitySerializer                  

class UserParticipationClubSerializer(serializers.Serializer):
    club = ClubSerializer()
    user = UserSerializer()

    class Meta:
        model = UserParticipationClub
        fields = "__all__"

class CreateUserParticipationClubSerializer(serializers.Serializer):
    user_id = serializers.UUIDField()
    club_id = serializers.UUIDField()

    def create(self, validated_data):
        user_id = validated_data.pop('user_id')
        club_id = validated_data.pop('club_id')
        return UserParticipationClub.objects.create(user_id=user_id, club_id=club_id, **validated_data)

    class Meta:
        model = UserParticipationClub
        fields = "__all__"


class UserParticipationEventSerializer(serializers.Serializer):
    # user = ActivitySerializer(many=False)
    event = EventSerializer()
    is_admin = serializers.BooleanField()
    is_superadmin = serializers.BooleanField()
    participated_at = serializers.DateTimeField()

    class Meta:
        model = UserParticipationEvent
        fields = (
            "is_admin",
            "is_superadmin",
            "participated_at",
            "user",
            "event"
        )

class CreateUserParticipationEventSerializer(serializers.Serializer):
    user_id = serializers.UUIDField()
    event_id = serializers.UUIDField()

    def create(self, validated_data):
        user_id = validated_data.pop('user_id')
        event_id = validated_data.pop('event_id')
        return UserParticipationEvent.objects.create(user_id=user_id, event_id=event_id, **validated_data)
    
    class Meta:
        model = UserParticipationEvent
        fields = "__all__"

class UserParticipationGroupSerializer(serializers.Serializer):
    group = GroupSerializer()
    user = UserSerializer()

    class Meta:
        model = UserParticipationGroup
        fields = "__all__"

class CreateUserParticipationGroupSerializer(serializers.Serializer):
    user_id = serializers.UUIDField()
    group_id = serializers.UUIDField()

    def create(self, validated_data):
        user_id = validated_data.pop('user_id')
        group_id = validated_data.pop('group_id')
        return UserParticipationGroup.objects.create(user_id=user_id, group_id=group_id, **validated_data)

    class Meta:
        model = UserParticipationGroup
        fields = "__all__"
